qualtran/bloqs/qrom.py

```diff
@@ -0,0 +1,87 @@
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from functools import cached_property
+from typing import Dict, Sequence, Tuple
+
+import cirq
+from attrs import frozen
+from cirq_ft import QROM as CirqQROM
+from numpy.typing import NDArray
+
+from qualtran import Bloq, CompositeBloq, Register, Signature
+from qualtran.cirq_interop import CirqQuregT, decompose_from_cirq_op
+
+
+@frozen
+class QROM(Bloq):
+    """Gate to load data[l] in the target register when the selection stores an index l.
+
+    In the case of multi-dimensional data[p,q,r,...] we use multiple named
+    selection registers [selection0, selection1, selection2, ...] to index and
+    load the data.
+
+    Args:
+        data: List of numpy ndarrays specifying the data to load. If the length
+            of this list is greater than one then we use the same selection indices
+            to load each dataset (for example, to load alt and keep data for
+            state preparation). Each data set is required to have the same
+            shape and to be of integer type.
+        selection_bitsizes: The number of bits used to represent each selection register
+            corresponding to the size of each dimension of the array. Should be
+            the same length as the shape of each of the datasets.
+        data_bitsizes: The number of bits used to represent the data
+            registers. This can be deduced from the maximum element of each of the
+            datasets. Should be of length len(data), i.e. the number of datasets.
+        cvs: The control values for the gate. Defaults to no controls.
+    """
+
+    data: Sequence[NDArray]
+    selection_bitsizes: Tuple[int, ...]
+    data_bitsizes: Tuple[int, ...]
+    cvs: Tuple[int, ...] = ()
+
+    @cached_property
+    def signature(self) -> Signature:
+        regs = [
+            Register(f"selection{i}", bitsize=bs) for i, bs in enumerate(self.selection_bitsizes)
+        ]
+        regs += [Register(f"target{i}", bitsize=bs) for i, bs in enumerate(self.data_bitsizes)]
+        if len(self.cvs) > 0:
+            regs += [Register("control", bitsize=1, shape=(len(self.cvs),))]
+        return Signature(regs)
+
+    def decompose_bloq(self) -> 'CompositeBloq':
+        return decompose_from_cirq_op(self)
+
+    def as_cirq_op(
+        self, qubit_manager: 'cirq.QubitManager', **cirq_quregs: 'CirqQuregT'
+    ) -> Tuple['cirq.Operation', Dict[str, 'CirqQuregT']]:
+        qrom = CirqQROM(
+            data=self.data,
+            selection_bitsizes=self.selection_bitsizes,
+            target_bitsizes=self.data_bitsizes,
+            num_controls=self.cvs,
```

Comment:
`CirqQROM`'s `num_controls` currently expects an `int` and not a `Tuple[int, ...]`, we should add support for arbitrary `cvs: Tuple[int, ...]` to the Cirq QROM but right now, this code should result in  an error. 

Can we add tests for the controlled version of `QROM` Bloq as well and fix the bug and raise an issue to track `cvs:Tuple[int, ...]` improvements to the Cirq / Qualtran QROM?

Reply:
@tanujkhattar should be done now!